File: quickstart_guides/trees/java/MaxDepthBinaryTree.java
/*
 *   Title: Max depth binary tree
 *
 * Problem:
 *    Given a binary tree, find height of it. Height of empty tree is 0 and
 *    height of below tree is 3.
 *
 *   Execution: javac MaxDepthBinaryTree.java && java MaxDepthBinaryTree
 */


class Node
{
    int data;
    Node left, right;

    Node(int item)
    {
        data = item;
        left = right = null;
    }
}

class MaxDepthBinaryTree
{
     Node root;

    /* Compute the "maxDepth" of a tree -- the number of
       nodes along the longest path from the root node
       down to the farthest leaf node.*/
    int maxDepth(Node node)
    {
        if (node == null)
            return 0;
        else
        {
            /* compute the depth of each subtree */
            int lDepth = maxDepth(node.left);
            int rDepth = maxDepth(node.right);

            /* use the larger one */
            if (lDepth > rDepth)
                return (lDepth + 1);
             else
                return (rDepth + 1);
        }
    }

    /* Driver program to test above functions */
    public static void main(String[] args)
    {
        BinaryTree tree = new BinaryTree();

        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);

        assert tree.maxDepth(tree.root) == 3;

        System.out.println("All tests passed.");
    }
}

File: quickstart_guides/trees/python/max_depth_binary_tree.py
"""
Title: Maximum depth of binary tree

Problem:
    Given a binary tree, find its maximum depth.

    The maximum depth is the number of nodes along the longest path from the
    root node down to the farthest leaf node.

    Note: A leaf is a node with no children.

Execution: python max_depth_binary_tree.py
"""
import unittest


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def max_depth_binary_tree(root: TreeNode) -> int:
    if root is None:
        return 0
    left = max_depth_binary_tree(root.left) + 1
    right = max_depth_binary_tree(root.right) + 1
    if left > right:
        return left
    return right


class TestMaxDepthBinaryTree(unittest.TestCase):
    """Unit tests for max_depth_binary_tree."""

    def test_1(self):
        tn = TreeNode(3)
        tn.left = TreeNode(9)
        tn.right = TreeNode(20)
        tn.right.left = TreeNode(15)
        tn.right.right = TreeNode(7)
        self.assertEqual(max_depth_binary_tree(tn), 3)


if __name__ == "__main__":
    unittest.main()
